service/src/main/kotlin/fi/espoo/evaka/pis/EmployeeQueries.kt

```diff
@@ -192,35 +192,67 @@ WHERE external_id = :externalId
 ).bind("externalId", externalId)
     .execute()
 
-fun Database.Transaction.updatePinCode(
+fun Database.Transaction.upsertPinCode(
     userId: UUID,
     pinCode: PinCode
 ) {
+    // Note: according to spec, setting a pin resets the failure and opens a locked pin
     // language=sql
     val sql = """
-        UPDATE employee
-        SET
-            pin = :pin_code
-        WHERE id = :userId
+INSERT INTO employee_pin(user_id, pin)
+VALUES(:userId, :pin)
+ON CONFLICT (user_id) DO UPDATE SET
+        user_id = :userId,
+        pin = :pin,
+        locked = false,
+        failure_count = 0
     """.trimIndent()
     val updated = this.createUpdate(sql)
         .bind("userId", userId)
-        .bind("pin_code", pinCode.pin)
+        .bind("pin", pinCode.pin)
         .execute()
 
-    if (updated == 0) throw NotFound("Could not update pin-code for $userId. User not found")
+    if (updated == 0) throw NotFound("Could not update pin code for $userId. User not found")
 }
 
 fun Handle.employeePinIsCorrect(employeeId: UUID, pin: String): Boolean = createQuery(
 """
 SELECT EXISTS (
     SELECT 1
-    FROM employee
-    WHERE id = :employeeId
+    FROM employee_pin
+    WHERE user_id = :employeeId
     AND pin = :pin
+    AND locked = false
 )
     """.trimIndent()
 ).bind("employeeId", employeeId)
     .bind("pin", pin)
     .mapTo<Boolean>()
     .first()
+
+fun Handle.resetEmployeePinFailureCount(employeeId: UUID) = createUpdate(
+    """
+UPDATE employee_pin
+SET failure_count = 0, locked = false
```

Comment:
I think you should remove the `locked = false` part here as well just in case since the locked state shouldn't be reset here.

Reply:
Apart from this thing this PR should be good to merge.